Performance: Cookie decoder improvements

jgoday opened the pull request:
See #571.

Reduce Cookie case class allocations in Cookie.decodeResponseCookie.
Right now, it allocates a new Cookie instance with each field (using Cookie.copy through several methods).
For example,
it allocates 3 instances for ```"SUID=123;httponly=true;expires=2007-12-03T10:15:30.00Z"```.

This change parses each field separately and then creates one single Cookie instance.

This simple jhm benchmark show the two approaches
```scala
package sample

import java.util.concurrent.TimeUnit
import org.openjdk.jmh.annotations._
import zhttp.http.Cookie

@State(Scope.Thread)
@BenchmarkMode(Array(Mode.Throughput))
@OutputTimeUnit(TimeUnit.MICROSECONDS)
class CookieDecode {
  @Benchmark
  def benchmarkDecode(): Unit = {
    val value="SUID=123;httponly=true;expires=2007-12-03T10:15:30.00Z"
    val _ = Cookie.decodeResponseCookie(value)
    ()
  }
}
```

Parsing and copying Cookie case class
```

# Warmup Iteration   1: 0,141 ops/us
[info] # Warmup Iteration   2: 0,151 ops/us
[info] # Warmup Iteration   3: 0,206 ops/us
[info] Iteration   1: 0,217 ops/us
[info] Iteration   2: 0,168 ops/us
[info] Iteration   3: 0,147 ops/us
[info] Result "sample.CookieDecode.benchmarkDecode":
[info]   0,177 ┬▒(99.9%) 0,653 ops/us [Average]
[info]   (min, avg, max) = (0,147, 0,177, 0,217), stdev = 0,036
[info]   CI (99.9%): [? 0, 0,831] (assumes normal distribution)
[info] # Run complete. Total time: 00:01:01
```

Parsing fields with mutable vars
```
[info] # Warmup Iteration   1: 0,340 ops/us
[info] # Warmup Iteration   2: 0,375 ops/us
[info] # Warmup Iteration   3: 0,354 ops/us
[info] Iteration   1: 0,369 ops/us
[info] Iteration   2: 0,375 ops/us
[info] Iteration   3: 0,349 ops/us
[info] Result "sample.CookieDecode.benchmarkDecode":
[info]   0,364 ┬▒(99.9%) 0,244 ops/us [Average]
[info]   (min, avg, max) = (0,349, 0,364, 0,375), stdev = 0,013
[info]   CI (99.9%): [0,120, 0,608] (assumes normal distribution)
[info] # Run complete. Total time: 00:01:00
```

jgoday committed:
#571: Reduce decodeResponseCookie allocations

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -147,25 +147,46 @@ object Cookie {
+        ),
+      )
+    } else {
+      None
```
This is great, you have almost doubled the perf! 

You can further reduce allocations by — 
1. Get rid of all the `Option` and use `null` instead. 
2. Every `map` operation is creating a new collection (more memory), use a for/while loop if possible
3. Get rid of interim `Function` and `Tuple` also.
4. Don't allocate the `String` inside the method, pull them outside so that thy are allocated only once.

NOTE: You can have a method called - `unsafeDecodeResponseCookie` that returns a null or a Cookie for performance and make it package private. Then also have a `decodeResponseCookie` which calls `unsafeDecodeResponseCookie` and wraps it in an `Option`

jgoday replied:
Made some changes to avoid string split and some allocations.

Marked internal functions _parseDate_ and _splitNameContent_ @inline functions, not sure if it could be a problem ...

tusharmath replied:
can you update the benchmark numbers as before and after. I think we can make more changes for performance :)

tusharmath replied:
Instead of using `None` use `null`. This can cause null pointer exceptions but that should be fine because we will rename this method to `unsafeDecodeResponseCookie` 
and make it package-private. And, we will have a separate method `decodeResponseCookie` which calls `unsafeDecodeResponseCookie` and wraps the output in an `Option`. And that  should be a public method.

jgoday committed:
avoid some allocations

jgoday committed:
use while instead of do while (scala 3 compatibility)

jgoday committed:
do not use string.isBlank

jgoday committed:
use string.isEmpty

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -143,30 +143,61 @@ object Cookie {
+      val c       = if (i <= 0) headerValue.substring(prev) else headerValue.substring(prev, i)
+      val (n, ct) = splitNameContent(c)
+
+      (n.toLowerCase, ct) match {
```
unnecessary tuple allocation.

jgoday replied:
@tusharmath I have mode some changes to avoid string and tuple unnecessary allocations, inner loop is not the fanciest but I think that this way you get more performance ... what do you think ?

tusharmath replied:
Can you commit the `Benchmark` code to the benchmarks project.

jgoday replied:
Sure, added CookieDecodeBenchark to zio-http-benchmarks (very naive at the moment)

jgoday committed:
Cookie decodeResponseCookie: avoid allocations

jgoday committed:
Add decode cookie benchmark code

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -176,24 +232,28 @@ object Cookie {
-      case _             => ("", None)
+  @inline
+  private def parseDate(v: String): Try[Instant] =
+    Try(Instant.parse(v))
```
Try will allocate additional memory. Instead, you could use a `try { } catch {}` block.
Incase of failure return a `null`

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -176,24 +232,28 @@ object Cookie {
+  private def splitNameContent(str: String): (String, String) = {
+    val i = str.indexOf('=')
+    if (i >= 0) {
+      (str.substring(0, i).trim, str.substring(i + 1).trim)
```
Extremely expensive computations. Every `substring` and `trim` has a `O(n)` space cmplx. If you can reduce this it will make a big difference.

tusharmath replied:
Here is a crazy idea :P 

What if you keep the original string as it is and at the time of parsing create a cookie that maintains indexes of start and stop on the original string. Thereby not allocating any intermediary memory at all. eg:
```scala
trait Cookie {
  private def raw: String
  private def nameStart: Int = ??? // fill this at the time of parsing
  private def nameEnd: Int   = ??? // fill this at the time of parsing
  def name: String = raw.substring(nameStart, nameEnd)
}
```
The code above is just to depict the idea, actual implementation might be very different. However, this would be dramatically faster than the original solution because you are only allocating when you need that data.

If you would like to do this index-based optimization in a different PR, I wouldn't mind that also :)

tusharmath commented on zio-http-benchmarks/src/main/scala/zhttp.benchmarks/CookieDecodeBenchmark.scala:
```diff
@@ -0,0 +1,19 @@
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class CookieDecodeBenchmark {
+  private val cookie = "SUID=123;httponly=true;expires=2007-12-03T10:15:30.00Z"
```
May be generate a larger cookie. With a large payload and more key-value pairs.

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -143,30 +143,86 @@ object Cookie {
+          } else {
+            name = headerValue.substring(0, next)
+          }
+        } else if (headerValue.regionMatches(true, curr, "expires=", 0, 8)) {
```
The string literals will also allocate memory every time the method is called. You could move it into a value so that it's allocated only once.

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -143,30 +143,86 @@ object Cookie {
+        } else if (headerValue.regionMatches(true, curr, "samesite=", 0, 9)) {
+          val v = headerValue.substring(curr + 9, next).toLowerCase
+          v match {
+            case "lax"    => sameSite = SameSite.Lax
```
I am not really sure but `"lax"` AFAIK will also allocate additional memory.

jgoday committed:
String literals and update CookieDecodeBenchmark

jgoday committed:
add unsafeDecodeResponseCookie

jgoday commented:
@tusharmath
Updated with your latest suggestions:
1. String literals
2. Update benchmark with a larger payload cookie
```
OLD CODE:
[info] CookieDecodeBenchmark.benchmarkApp  thrpt    3  387562,885 ± 52538,364  ops/s

NEW CODE:
[info] CookieDecodeBenchmark.benchmarkApp  thrpt    3  679569,604 ± 27421,427  ops/s
```
3. Avoid scala.util.Try in parseDate
4. Added unsafeDecodeResponseCookie: Cookie
5. Really like the idea of having a cookie with the index-based fields, specially to avoid splitNameContent allocations in 'decodeRequestCookie', but as you said, maybe it's better to do it in another PR ...

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -176,24 +246,24 @@ object Cookie {
-      case Array(v1, v2) => (v1, Some(v2))
-      case _             => ("", None)
+  @inline
+  private def splitNameContent(str: String): (String, String) = {
```
Avoid tuple creation.

tusharmath replied:
You can create two separate functions that give you different parts viz. name and value, from the string.

jgoday replied:
splitNameContent is currently only used in decodeRequestCookie, to avoid this tuple allocation we can just chante it to
```scala
val name    =
  if (i >= 0) a.substring(0, i).trim
  else a.trim
val content =
  if (i >= 0) a.substring(i + 1).trim
  else null

if (name.isEmpty && content.isEmpty) None
else Some(Cookie(name, content))
```

however, maybe this is the best candidate to use a index base view, like  you mentioned earlier ...
```
case class CookieView(cookieStr) implements Cookie {
  def name: String = ...
  def content: String = ....
}
```

tusharmath commented on zio-http-benchmarks/src/main/scala/zhttp.benchmarks/CookieDecodeBenchmark.scala:
```diff
@@ -0,0 +1,31 @@
+    true,
+    Some(1200),
+    Some(Cookie.SameSite.Strict),
+  )
```
Can we use a strings that are at least 100 chars or more. 

```scala
val random = new scala.util.Random((Math.random * Int.MaxValue).toInt)
val name = random.alphanumeric.take(100).mkString("")
val value = random.alphanumeric.take(100).mkString("")
val domain = random.alphanumeric.take(100).mkString("")
val name = Path((0 to 10).map { _ => random.alphanumeric.take(10).mkString("") })
```

The real improvement in performance will be observed in larger strings.

jgoday replied:
Updated benchmark code with the following results:
```
Old code:
(min, avg, max) = (393771,450, 395268,674, 396456,159), stdev = 1368,893

New code:
(min, avg, max) = (634553,745, 636047,487, 637677,233), stdev = 1566,179

```

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -176,24 +246,24 @@ object Cookie {
     val cookies: Array[String]  = headerValue.split(";").map(_.trim)
```
```suggestion
    val cookies: Array[String]  = headerValue.split(';').map(_.trim)
```

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -176,24 +246,24 @@ object Cookie {
       val (name, content) = splitNameContent(a)
-      if (name.trim == "" && content.isEmpty) None
-      else Some(Cookie(name, content.getOrElse("")))
+      if (name == "" && content.isEmpty) None
```
```suggestion
      if (name.isEmpty && content.isEmpty) None
```

jgoday committed:
Update zio-http/src/main/scala/zhttp/http/Cookie.scala

Co-authored-by: Tushar Mathur <tusharmath@gmail.com>

jgoday committed:
Update zio-http/src/main/scala/zhttp/http/Cookie.scala

Co-authored-by: Tushar Mathur <tusharmath@gmail.com>

jgoday committed:
Update cookie decode benchmark

tusharmath commented on zio-http/src/main/scala/zhttp/http/Cookie.scala:
```diff
@@ -142,58 +153,117 @@ object Cookie {
+            try { Instant.parse(headerValue.substring(curr + 8, next)) }
+            catch { case _: Throwable => null }
+        } else if (headerValue.regionMatches(true, curr, fieldMaxAge, 0, fieldMaxAge.length)) {
+          maxAge = Try(headerValue.substring(curr + 8, next).toLong).toOption
```
We can create a global `try` `catch` block for this method.

jgoday replied:
@tusharmath , a couple of questions here:
1) would it be enough to wrap with a try each attempt to decode a field?

2) What should we do with the exception, discard it completely?

tusharmath replied:
1. We should have only 1 top level try catch block inside the `safe` implementation.
2. Incase of failure we should throw an exception in the `unsafe` implementation. The safe implementation captures it and then could do either of the two things —
   a. Ignore the error and return a `None`.
   b. Capture the error and return a `Left[Throwable]`

jgoday replied:
Ok, changed to catch the exception in decodeResponseCookie but keeping the Option[Cookie] signature

tusharmath commented:
Really looking forward to this PR @jgoday :)

jgoday committed:
do not catch exceptions in unsafeDecodeResponseCookie

tusharmath commented on zio-http-benchmarks/src/main/scala/zhttp.benchmarks/CookieDecodeBenchmark.scala:
```diff
@@ -0,0 +1,38 @@
+
+  @Benchmark
+  def benchmarkApp(): Unit = {
+    val _ = Cookie.decodeResponseCookie(cookieStr)
```
```suggestion
    val _ = Cookie.unsafeDecodeResponseCookie(cookieStr)
```

tusharmath replied:
Can you run this one last time and post the results :)

jgoday replied:
Ok, applied your suggestionts.
I have run the benchmark again, here are the results

jmh:run -i 3 -wi 3 -f1 -t4

Old code:
```
[info] Iteration   1: 379730,343 ops/s
[info] Iteration   2: 383230,774 ops/s
[info] Iteration   3: 381724,031 ops/s
[info] Result "sample.CookieDecodeBenchmark.benchmarkApp":
[info]   381561,716 ±(99.9%) 32033,267 ops/s [Average]
[info]   (min, avg, max) = (379730,343, 381561,716, 383230,774), stdev = 1755,851
[info]   CI (99.9%): [349528,449, 413594,983] (assumes normal distribution)
[info] # Run complete. Total time: 00:01:00
```

New code:
```
[info] Iteration   1: 653978,242 ops/s
[info] Iteration   2: 635232,010 ops/s
[info] Iteration   3: 633734,049 ops/s
[info] Result "sample.CookieDecodeBenchmark.benchmarkApp":
[info]   640981,434 ±(99.9%) 205797,553 ops/s [Average]
[info]   (min, avg, max) = (633734,049, 640981,434, 653978,242), stdev = 11280,458
[info]   CI (99.9%): [435183,881, 846778,986] (assumes normal distribution)
[info] # Run complete. Total time: 00:01:00

```

jgoday committed:
CookieDecodeBenchmark agains unsafeDecodeResponseCookie

tusharmath commented:
@jgoday Thanks! this was pretty amazing :) We make it 1.7x faster. It might not seem much right now, but it all adds up when write a proper e2e application.

jgoday commented:
Thank you @tusharmath  for all your help and support.
